fusion.md

```diff
@@ -0,0 +1,1396 @@
+ *          GCAM_PARALLEL_ENABLED the appropriate "scratch" space to reset is identified
+ *          as the one assigned to the calling thread via the thread local Value::sCentralValue.
+ */
+void ManageStateVariables::copyState() {
+#if !GCAM_PARALLEL_ENABLED
+    memcpy( mStateData[1], mStateData[0], (sizeof( double)) * mNumCollected );
+#else
+    memcpy( Value::sCentralValue.local(), mStateData[0], (sizeof( double)) * mNumCollected );
+#endif
+}
+```
+
+
+
+Note that with GCAM parallel `Value::sCentralValue` is a thread local variable thus each variable can be indpendently set by each thread that is accessing that code.  What this means in practical terms is for instance that the electricity technology Gas CC could have calculated different costs at the same exact time depending on which computation thread is asking.
+
+
+
+Once we are done solving the period the `ManageStateVariables` will loop back over each state Value and reset the `mIsStateCopy` flag and copy back the "base" state value for long term storage.  Also releasing the memory for the centrally managed state's arrays.
+
+
+
+#### Ensuring that no state variable is missed
+
+
+We can check to make sure that not Data definitions were missed being tagged as
+"state" by enabling the preprocessor flag `DEBUG_STATE` which will enable checks
+to flag Values that are changed during a call to `World::calc` as well as other
+checks to ensure Values get collected / reset properly.
+
+
+
+Other GCAM Fusion related utilities
+-----------------------------------
+
+
+
+### ExpandDataVector
+
+Generally developers will not need to call this method directly.  Instead, it is
+used indirectly through searches via GCAM Fusion.  It is a utility for ensuring
+that we get the complete data vector from a data container taking into account
+the data vectors inherited from any base classes.  Expanding the full data
+vector is more tricky than it would first appear since we need to be able to
+determine which SubClass we are dealing with at runtime as we only ever store
+instances with the Base class pointer (this is typically accomplished with
+virtual methods). However the return type of each SubClass would be different
+for each SubClass. Thus we need to use a double dispatch based approach with a
+visitor that will collect the full data vector. In order for this visitor to be
+generic it needs to be templated however mixing virtual methods with templated
+argument is not allowed by the compiler due to possibly infinite method
+combinations.
+
+
+
+### Factory
+
+A generic templated factory that can create any member of a
+[SubClassFamilyVector](#definedata---and-definedatawithparent--) given the XML
+name.  This class is currently not used however could be employed to replace all
+of the various Factory singleton classes that currently exist in GCAM.  It would
+really be useful when/if we generate all XML Parsing code by the compiler.
+
+
+
+C++11/14 Features:
```

Comment:
I'm not sure how useful the material from here on is for the documentation, since it doesn't really pertain to GCAM per se.  Maybe another observer could weigh in.  @kvcalvin ?